### PR TITLE
Add support for custom TODO format

### PR DESCRIPTION
**Pull request by sattellite**
This PR related to #53

### CODE REVIEW

**Comment by preslavmihaylov** [comment]
Looks good so far. 

Thanks for taking this on, I think it will be a bumpy issue to execute, but very valuable and much appreciated if you manage to pull it off!

Some pointers - you can look at the `matchers` package and make the regexes in e.g. `matchers/scripts/todomatcher.go` to accept the TODO as a parameter.

To avoid making too many if-else statements, meant to preserve the current behaviour, you can make the config accept a default value for `custom_todos` as `[]string{"TODO"}` and always rely on this field. That way, you won't have to make checks such as "if this field is present, use behaviour A, otherwise, use behaviour B"

**Comment by preslavmihaylov** [comment]
Also, the default `TODO` format should always be supported, regardless of the field being present or not.
Additionally, I would love to see some integration tests in the `testing` package in the end to see this in action.

File: config/config.go
Comment (preslavmihaylov) [inline config/config.go:162]: The name `checkTodos` doesn't really review the func's intent. One has to read the docstring or implementation to understand its purpose.

I suggest renaming this to something more understandable such as `addDefaultFormatIfMissing`

Comment (sattellite) [reply config/config.go]: Renamed

**Comment by sattellite** [comment]
I only have to add tests for this functionality.

**Comment by preslavmihaylov** [comment]
thanks @sattellite I'll spend some time to review this during the weekend as it's a rather hard week for me so far. :(

File: matchers/matchers.go
Comment (preslavmihaylov) [inline matchers/matchers.go:170]: This is not the place for such a function.

The problem is that you are writing it with the knowledge that all underlying todo matchers operate with regex.
This is an implementation detail and if at some point, we decide to migrate from regex to something else, this function here will cause a lot of issues.

Instead, this function can be defined in the `common` package as `arrayAsRegexAnyMatchExpression`.
The function declaration should instead be `[]string` and the specifics of turning the array of strings into a regex expression can be done for every todo matcher.

File: matchers/php/todomatcher.go
Comment (preslavmihaylov) [inline matchers/php/todomatcher.go:19]: I guess these global variables are used so that we don't initialise all these regexps every time a new matcher for a file is created, which happens for every file.

Instead of making this optimization here, you can do it in the matcher factories. There, instead of initializing a matcher every time, initialise them as fields of the `matcherFactory` and reuse the instances.

If we implement it with these global variables, it would lead to one having to understanding the implementation details of this component to use them properly.

Given the folllowing snippet:
```
matcherA = php.NewTodomatcher("TODO", "todo")
matcherB = php.NewTodomatcher("FIXME", "fixme")
```

`matcherA` will now work with `FIXME` and `fixme` instead of the formats it was initialised with.

File: matchers/php/todomatcher.go
Comment (preslavmihaylov) [inline matchers/php/todomatcher.go:24]: There is no need for using this once we remove the global variables.

File: matchers/php/todomatcher.go
Comment (preslavmihaylov) [inline matchers/php/todomatcher.go:48]: Once we remove the global variables, all these can be condensed into a single return statement, which creates the struct.

File: matchers/python/todomatcher.go
Comment (preslavmihaylov) [inline matchers/python/todomatcher.go:2]: The comment on the previous todo matcher applies for all of them.

Not going to repeat the same stuff all over the place.

File: testing/scenarios/custom_todos/standart/main.go
Comment (preslavmihaylov) [inline testing/scenarios/custom_todos/standart/main.go:1]: The directory should be `standard` not `standart`

File: matchers/matchers.go
Comment (preslavmihaylov) [inline matchers/matchers.go:133]: Instead of creating a `Matchers` struct here just to hold the custom todos, you can add an additional field in the `checker` and pass along the custom todos on each call to `TodoMatcherForFile `

**Review by preslavmihaylov — CHANGES_REQUESTED** [review]
I haven't commented on absolutely everything, but there's plenty to work on still.

After the next iteration of changes, I'll go over the PR again with additional comments.

However, you're moving in the right direction.

File: config/config.go
Comment (sattellite) [reply config/config.go]: Added function to decode escaped strings. I want add tests for it but I don't understand where to place and how to name the  tests for it.

File: testing/scenarios/custom_todos/standart/main.go
Comment (sattellite) [reply testing/scenarios/custom_todos/standart/main.go]: Renamed the directory.

File: matchers/php/todomatcher.go
Comment (sattellite) [reply matchers/php/todomatcher.go:19]: I don't understand what you mean with this words. I have never seen or done this before. Can you give me links for read about this? Or describe little more please. This is my first language with strict types and I have never worked with factories before. So I don't understand how to transfer this optimization from matchers to the factory. I would be very grateful for your help.

Comment (sattellite) [reply matchers/php/todomatcher.go:19]: > Instead of making this optimization here, you can do it in the matcher factories. There, instead of initializing a matcher every time, initialise them as fields of the `matcherFactory` and reuse the instances.

I thought for all night about the problem described and I think I understood what you wrote about. Is it a correct solution?

```go
var lock sync.Mutex

type matcherFactory struct {
	newTodoMatcher     func(string) TodoMatcher
	newCommentsMatcher func(callback state.CommentCallback) CommentMatcher
	todoInstance       TodoMatcher
}

var standardMatcherFactory = &matcherFactory{
	func(customTodos string) TodoMatcher {
		if standardMatcherFactory.todoInstance == nil {
			lock.Lock()
			standardMatcherFactory.todoInstance = standard.NewTodoMatcher(customTodos)
			lock.Unlock()
		}
		return standardMatcherFactory.todoInstance
	},
	func(callback state.CommentCallback) CommentMatcher {
		return standard.NewCommentMatcher(callback, false)
	},
}
```

File: matchers/matchers.go
Comment (sattellite) [reply matchers/matchers.go]: Fixed

File: matchers/matchers.go
Comment (sattellite) [reply matchers/matchers.go]: Renamed and moved function to `common`

File: matchers/php/todomatcher.go
Comment (preslavmihaylov) [reply matchers/php/todomatcher.go:20]: yes, like this but no need for the lazy initialization.

Simply initialise it like so:
```go
var standardMatcherFactory = &matcherFactory{
        todoMatcher: standard.NewTodoMatcher()
	func(customTodos string) TodoMatcher {
		return standardMatcherFactory.todoMatcher
	},
	func(callback state.CommentCallback) CommentMatcher {
		...
	},
}
```

The `customTodos` need not be provided here. Provide them in the instance functions of the todo matcher. This avoids creating a state variable.

Comment (sattellite) [reply matchers/php/todomatcher.go:20]: Sorry in advance for stupid questions and inferences. I guess I don't understand what it should look like at all.

> The `customTodos` need not be provided here. Provide them in the instance functions of the todo matcher. This avoids creating a state variable.

Is this mean that I need add function like `setTodos` for matchers? When this function must be called?

I don't understand why this change will make the code clearer. There will be state variables with regular expressions in any matcher. This will make it possible to change sets of regular expressions at runtime. The configuration is set once at startup and is not changed at runtime, so I don't see the point in this method.

Comment (preslavmihaylov) [reply matchers/php/todomatcher.go:20]: hmm, okay. You're right that the matchers will have the regexps as state variables & if we provide the `customTodos` on each call to `IsMatch` and `IsValid`, then the regular expressions will need to be reinitialized.

In this case, this approach is ok with a suggestion for improving the synchronization like this:
```go
var (
	once            *sync.Once
	standardMatcher TodoMatcher
)

var (
	standardMatcherFactory = &matcherFactory{
		func(customTodos string) TodoMatcher {
			once.Do(func() {
				standardMatcher = standard.NewTodoMatcher(customTodos)
			})

			return standardMatcher
		},
		func(callback state.CommentCallback) CommentMatcher {
			return standard.NewCommentMatcher(callback, false)
		},
	}
        ...
```

Comment (sattellite) [reply matchers/php/todomatcher.go]: Optimized. Thank you for your advice.

File: testing/todocheck_test.go
Comment (preslavmihaylov) [inline testing/todocheck_test.go:806]: Could you extract all new tests from this file into a `customtodos_test.go`?

Just to keep files small

File: traverser/todoerrs/todoerrs.go
Comment (preslavmihaylov) [inline traverser/todoerrs/todoerrs.go:20]: I feel like `todos` is ambiguous here, especially given that it isn't an explicit type, it is a `[]string`.

Suggest renaming it to `customTodos` for clarity

**Review by preslavmihaylov — CHANGES_REQUESTED** [review]
This is almost complete now.

I have several minor issues to address still, after which we are finally ready to merge.

Thanks for all your efforts, this was a big one!

File: matchers/matchers.go
Comment (sattellite) [reply matchers/matchers.go:51]: `standardMatcher` used in two places: `standardMatcherFactory` and `standardMatcherWithNestedMultilineCommentsFactory`. This check will avoid second initialization of `standardMatcher`

**Comment by sattellite** [comment]
Please check README. English is not my native language and I often make mistakes in it.

**Review by preslavmihaylov — CHANGES_REQUESTED** [review]

File: matchers/matchers.go
Comment (sattellite) [reply matchers/matchers.go]: You are was right. I wrote test for multiple matchers and it was failed. I rewrote code with your advice with changed `once` scope and memoized `matcher` and test passed.

File: matchers/matchers.go
Comment (preslavmihaylov) [inline matchers/matchers.go:71]: Why are these inside the returned function instead of being in the closure, similar to how you've done it for `standard` matchers?

**Review by preslavmihaylov — CHANGES_REQUESTED** [review]

File: matchers/matchers.go
Comment (sattellite) [reply matchers/matchers.go]: Apparently, I was distracted. I corrected this misprint.

File: README.md
Comment (preslavmihaylov) [inline README.md:247]: Let's rephrase it a bit to make the target use-case more evident:
> By default, `todocheck` looks for todos in the format `// TODO #231: ...`. Most projects stick to this format. 
> 
> However, if you have some special todo format, which is different from the default one, you can specify a > list of custom todo formats in your `.todocheck.yaml`

File: README.md
Comment (preslavmihaylov) [inline README.md:256]: > Note that this list adds additional formats the tool should match, but doesn't substitute the default one - `TODO`.

File: README.md
Comment (preslavmihaylov) [inline README.md:255]: Let's also specify an example usage:

With the above configuration, `todocheck` will not start matching todo comments in the following formats:
```
// @fix #234: a leftover task...
// ToDo #15: A funky-looking todo comment
```

**Review by preslavmihaylov — CHANGES_REQUESTED** [review]

**Comment by preslavmihaylov** [comment]
Wow, we've come a long way with this PR. Thank you tremendously for your efforts!

Really looking forward to shipping this soon.

**Comment by sattellite** [comment]
Thank you for spending so much time checking my code. You are a great software architect. And you are a very cool reviewer. This pull request was very useful for me thanks to your experience and efforts.

**Review by preslavmihaylov — APPROVED** [review]
I am very happy to finally approve these changes & hit the merge button!

Congratulations on such a significant contribution 🎉

**Pull request merged**
